### PR TITLE
ddl: add syntax for setting the cache step of auto id explicitly.

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?
That autoid allocator's cache step is quite big now, it will consume a lot of autoid when tidb restarts or crashes. For those scenes which treasure autoid heavily, it's not acceptable.

Parser link: [#765](https://github.com/pingcap/parser/pull/765)

Problem Summary:

### What is changed and how it works?
Add `auto_increment_cache` in create table / alter table statement and it will determine the allocator's cache step.

**But there is an exception when a statement like an insert batch(`insert into values()()...`) which requires allocating consecutive N autoid in one statement, we will make sure that the custom cache step is adequate for it with `if step < N, then step = min(2N, maxstep)`**

>Attention please: 
tidb handle share the same allocator with auto_increment, so here:
`auto_increment_cache` will take effect on `handle` if `PKIshandle` is `false`, or even there is no auto_increment column at all. 

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note: 
ddl: add syntax for setting the cache step of auto id explicitly.

### CODE REVIEW

**Comment by AilinKid** [comment]
This PR need parser merged first

File: executor/show.go [hunk 1]
```diff
@@ -886,6 +886,10 @@ func ConstructResultOfShowCreateTable(ctx sessionctx.Context, tableInfo *model.T
 		}
 	}
 
+	if tableInfo.AutoIdCache != 0 {
+		fmt.Fprintf(buf, " AUTO_ID_CACHE=%d", tableInfo.AutoIdCache)
```
Comment (Deardrops) [inline executor/show.go]: How about an output like ` /*!90000 AUTO_ID_CACHE=%d */`

File: meta/autoid/autoid.go [hunk 2]
```diff
@@ -375,7 +394,11 @@ func NewSequenceAllocator(store kv.Storage, dbID int64, info *model.SequenceInfo
 func NewAllocatorsFromTblInfo(store kv.Storage, schemaID int64, tblInfo *model.TableInfo) Allocators {
 	var allocs []Allocator
 	dbID := tblInfo.GetDBID(schemaID)
-	allocs = append(allocs, NewAllocator(store, dbID, tblInfo.IsAutoIncColUnsigned(), RowIDAllocType))
+	if tblInfo.AutoIdCache > 0 {
```
Comment (Deardrops) [inline meta/autoid/autoid.go]: Do there a test for the case `AutoIdCache == -10`?

File: executor/show.go (hunk 1 above)
Comment (AilinKid) [reply executor/show.go]: nice idea.

File: meta/autoid/autoid.go (hunk 2 above)
Comment (AilinKid) [reply meta/autoid/autoid.go]: this element should be `GE` to 0 (default), otherwise, create table statement will parser error.

**Comment by bb7133** [comment]
LGTM

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -3143,6 +3147,33 @@ func (d *ddl) AlterTableComment(ctx sessionctx.Context, ident ast.Ident, spec *a
...
+		return errors.Trace(infoschema.ErrTableNotExists.GenWithStackByArgs(ident.Schema, ident.Name))
+	}
+
+	job := &model.Job{
+		SchemaID:   schema.ID,
+		TableID:    tb.Meta().ID,
+		SchemaName: schema.Name.L,
+		Type:       model.ActionModifyTableAutoIdCache,
+		BinlogInfo: &model.HistoryInfo{},
+		Args:       []interface{}{newCache},
```
Comment (Deardrops) [inline ddl/ddl_api.go:3169]: Do we need check `newCache` is 0 or overflowed before start DDL job?

Comment (AilinKid) [reply ddl/ddl_api.go:3177]: addressed, `newCache=0` is allowed, meaning cancel the user-defined auto_id_cache.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -3143,6 +3155,33 @@ func (d *ddl) AlterTableComment(ctx sessionctx.Context, ident ast.Ident, spec *a
...
+	schema, ok := is.SchemaByName(ident.Schema)
+	if !ok {
+		return infoschema.ErrDatabaseNotExists.GenWithStackByArgs(ident.Schema)
+	}
+
+	tb, err := is.TableByName(ident.Schema, ident.Name)
+	if err != nil {
+		return errors.Trace(infoschema.ErrTableNotExists.GenWithStackByArgs(ident.Schema, ident.Name))
+	}
+
```
Comment (djshow832) [inline ddl/ddl_api.go:3170]: What about the tables that don't have auto_increment column?

Comment (AilinKid) [reply ddl/ddl_api.go:3170]: showed in description

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -3143,6 +3155,33 @@ func (d *ddl) AlterTableComment(ctx sessionctx.Context, ident ast.Ident, spec *a
...
+	is := d.infoHandle.Get()
+	schema, ok := is.SchemaByName(ident.Schema)
+	if !ok {
+		return infoschema.ErrDatabaseNotExists.GenWithStackByArgs(ident.Schema)
+	}
+
+	tb, err := is.TableByName(ident.Schema, ident.Name)
+	if err != nil {
+		return errors.Trace(infoschema.ErrTableNotExists.GenWithStackByArgs(ident.Schema, ident.Name))
+	}
```
Comment (zimulala) [inline ddl/ddl_api.go:3169]: Could we use `getSchemaAndTableByIdent` instead of line3160-3169?

File: executor/show.go [hunk 6]
```diff
@@ -900,6 +900,10 @@ func ConstructResultOfShowCreateTable(ctx sessionctx.Context, tableInfo *model.T
 		}
 	}
 
+	if tableInfo.AutoIdCache != 0 {
+		fmt.Fprintf(buf, " /*!90000 AUTO_ID_CACHE=%d */", tableInfo.AutoIdCache)
```
Comment (zimulala) [inline executor/show.go:904]: Please add a test for this logic.

File: ddl/ddl_api.go (hunk 5 above)
Comment (AilinKid) [reply ddl/ddl_api.go:3414]: addressed

File: ddl/db_integration_test.go [hunk 7]
```diff
@@ -2152,3 +2153,108 @@ func (s *testIntegrationSuite7) TestAddExpressionIndexOnPartition(c *C) {
...
+	tk.MustExec("drop table if exists t;")
+	_, err = tk.Exec("create table t(a int) auto_id_cache = 9223372036854775808")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "table option auto_id_cache overflows int64")
+
+	tk.MustExec("create table t(a int) auto_id_cache = 9223372036854775807")
+	_, err = tk.Exec("alter table t auto_id_cache = 9223372036854775808")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "table option auto_id_cache overflows int64")
+
```
Comment (zimulala) [inline ddl/db_integration_test.go:2259]: Please remove the useless lint.

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -1923,6 +1923,12 @@ func handleTableOptions(options []*ast.TableOption, tbInfo *model.TableInfo) err
 		switch op.Tp {
 		case ast.TableOptionAutoIncrement:
 			tbInfo.AutoIncID = int64(op.UintValue)
+		case ast.TableOptionAutoIdCache:
+			if op.UintValue > uint64(math.MaxInt64) {
+				// Todo: refine this error.
```
Comment (zimulala) [inline ddl/ddl_api.go:1928]:
```suggestion
				// TODO: Refine this error.
```

File: ddl/ddl_api.go [hunk 9]
```diff
@@ -2150,6 +2156,12 @@ func (d *ddl) AlterTable(ctx sessionctx.Context, ident ast.Ident, specs []*ast.A
 					err = d.ShardRowID(ctx, ident, opt.UintValue)
 				case ast.TableOptionAutoIncrement:
 					err = d.RebaseAutoID(ctx, ident, int64(opt.UintValue))
+				case ast.TableOptionAutoIdCache:
+					if opt.UintValue > uint64(math.MaxInt64) {
+						// Todo: refine this error.
```
Comment (zimulala) [inline ddl/ddl_api.go:2161]: Ditto.

**Pull request merged**